Part II Config enhancements

tmgstevens opened the pull request:
This change further improves the idempotency of the redpanda_broker role by allowing for multi-modal operation, supporting both first run (via ``.bootstrap.yaml`) and nth run scenarios (via `rpk cluster config set` commands).

Refactored the TLS playbooks to take the redpanda specific configuration out of the playbooks and simple to re-run the `redpanda_broker` role (with `tls: true`) once certificates have been installed.

rkruze commented on ansible/playbooks/roles/redpanda_broker/templates/configs/defaults.j2:
```diff
@@ -1,31 +1,49 @@
+  cluster_id: "{{ redpanda_cluster_id }}"
   redpanda:
+    data_directory: "{{ redpanda_data_directory }}"
     node_id: "{{ id }}"
```
Should we remove node_id?

WesWWagner replied:
I just manually removed node id and did a full reboot of all nodes and the cluster functions as expected.

WesWWagner replied:
I recommend that we remove all elements of node+id as part of this PR to make adding and removing nodes cleaner.

tmgstevens replied:
Removing node id is part of #95 . I'm not sure about the edge cases with seed servers yet

tmgstevens replied:
I've removed `node_id` - seems to work okay

WesWWagner commented:
Flattening of the TLS playbooks into a single directory results in README.md having incorrect path references.

WesWWagner commented:
validated that while adding 2 nodes, TLS did not overwrite the existing TLS configs on existing nodes. 

We will need a "force-tls-refresh" or something similar in the future to support actual cert renewals and reissuance.

PLAY RECAP **************************************************************************************************************************
34.215.102.124             : ok=9    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   
54.185.93.24               : ok=9    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   
54.190.109.94              : ok=9    changed=7    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   
54.200.115.253             : ok=9    changed=7    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   
54.70.189.179              : ok=9    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0

WesWWagner commented:
Running install-certs throws allegedly FATAL errors on existing nodes when installing the CERTS on the new nodes:
TASK [redpanda_broker : start redpanda] *********************************************************************************************
fatal: [34.215.102.124]: FAILED! => {"changed": false, "msg": "Unable to start service redpanda: Job for redpanda.service failed because the control process exited with error code.\nSee \"systemctl status redpanda.service\" and \"journalctl -xe\" for details.\n"}
fatal: [54.70.189.179]: FAILED! => {"changed": false, "msg": "Unable to start service redpanda: Job for redpanda.service failed because the control process exited with error code.\nSee \"systemctl status redpanda.service\" and \"journalctl -xe\" for details.\n"}
fatal: [54.185.93.24]: FAILED! => {"changed": false, "msg": "Unable to start service redpanda: Job for redpanda.service failed because the control process exited with error code.\nSee \"systemctl status redpanda.service\" and \"journalctl -xe\" for details.\n"}

Start redpanda fails for already running nodes, but the cluster is left in a correct running state.

tmgstevens commented:
> Running install-certs throws allegedly FATAL errors on existing nodes when installing the CERTS on the new nodes: TASK [redpanda_broker : start redpanda] ********************************************************************************************* fatal: [34.215.102.124]: FAILED! => {"changed": false, "msg": "Unable to start service redpanda: Job for redpanda.service failed because the control process exited with error code.\nSee "systemctl status redpanda.service" and "journalctl -xe" for details.\n"} fatal: [54.70.189.179]: FAILED! => {"changed": false, "msg": "Unable to start service redpanda: Job for redpanda.service failed because the control process exited with error code.\nSee "systemctl status redpanda.service" and "journalctl -xe" for details.\n"} fatal: [54.185.93.24]: FAILED! => {"changed": false, "msg": "Unable to start service redpanda: Job for redpanda.service failed because the control process exited with error code.\nSee "systemctl status redpanda.service" and "journalctl -xe" for details.\n"}
> 
> Start redpanda fails for already running nodes, but the cluster is left in a correct running state.

Can you get me the logs why it wouldn't start please? I'll look tomorrow

WesWWagner commented:
When adding a node to a cluster, provision_node does a restart of all nodes ungracefully. If you attempt a limit <IP OF HOST> there are errors in configuration of the new node.

There is a similar pitfall in install_certs.yml where all nodes in scope are restarted.

tmgstevens commented:
Wes - the restart is throttled to 1 so shouldn't do more than one at a
time. It will cause some leadership election, but realistically I wouldn't
expect people to be doing this on a live system anyhow.

What errors are you referring to? I've not seen that behaviour. Please
could you provide some Ansible logs and Redpanda logs so that I can review?

On Tue, 22 Nov 2022, 21:31 Wes Wagner, ***@***.***> wrote:

> When adding a node to a cluster, provision_node does a restart of all
> nodes ungracefully. If you attempt a limit there are errors in
> configuration of the new node.
>
> There is a similar pitfall in install_certs.yml where all nodes in scope
> are restarted.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/deployment-automation/pull/96#issuecomment-1324260364>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIIEFG3S7RL7YCETGWYCL3WJU3TLANCNFSM6AAAAAASDMTVIQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

WesWWagner commented:
These changes need to be made to documentation to make this code workable as-is:
1) when running install-certs you need to set extra-vars to enable node restarting:
--extra-vars "restart_node=true"
2) after enabling TLS is is imperative that you add tls=true entries to your hosts in the ini file. We should probably automate that somehow.
3) when adding a node with the terraform scripts, it will overwrite your previous tls=true entries, so you have a good chance of accidentally getting the cluster into a bad state when using the AWS terraform scripts
Here is what a hosts.ini should look like when you are adding a new node and trying to enable TLS:
[redpanda]
35.164.216.186 ansible_user=ubuntu ansible_become=True private_ip=172.31.8.225 id=0 tls=true
35.90.138.83 ansible_user=ubuntu ansible_become=True private_ip=172.31.5.239 id=1 tls=true
18.236.236.214 ansible_user=ubuntu ansible_become=True private_ip=172.31.2.154 id=2 tls=true
54.191.210.208 ansible_user=ubuntu ansible_become=True private_ip=172.31.13.206 id=3 restart_node=true
4) If adding a TLS enabled node to the cluster it is best done through the install-certs playbook so it does not bring it up in a non tls state first before then enabling tls with a subsequent run. When you do this you cannot limit the execution to just the new node because it needs facts from the other servers. That is where the tls=true and the restart_node=true settings in the ini file are critical.

TLS enablement breaks the [generate the redpanda grafana dashboard] TASK when running a normal node install --- this is an issue in the existing codebase not related to this PR and should not hold up this PR.

tmgstevens commented:
@WesWWagner I've re-tested again yesterday and today and made a few changes including documenting the slightly non-obvious stuff.

Current status:
- Set `restart_node` back to `true` by default because actually setting it to `false` can lead you in a much worse situation. Added readme comment to this effect.
- Added `skip_node` which defaults to `false` - you can use this to completely skip applying the redpada_broker node to specific nodes.
- Reworked the way that the advertised_ip of _other_ nodes is calculated because this failed when they were skipped.
- Comprehensively documented the process for adding new nodes (with and without TLS) and also the process for creating a cluster with TLS enabled in one shot.
- Documented all of these extra parameters.
- Removed `node_id`.

I think this is good to go now. I'll create an issue for implementing proper rolling restarts via maintenance mode and we can fix up the grafana stuff later on as well. If you're happy, let's get this approved and merged.

WesWWagner commented:
Running install-certs to try to add a node to a new cluster for the first time results in this error:

fatal: [18.236.225.246]: FAILED! => {"changed": false, "checksum": "e1bce48f12275b6e36dfd3f0adec752a2b895e81", "msg": "Destination directory /etc/redpanda/certs does not exist"}

It would be unwise to run the standard node installation script first before applying TLS because you would have a middle period where the cluster is part TLS part not TLS and partition leaders could get migrated and the client would not negotiate properly.

WesWWagner commented:
Following this path:

Attempting to add a node w/ TLS enabled...
Add the new host(s) to the hosts.ini file. You may add skip_node=true to the existing hosts to avoid the playbooks being re-run on those nodes.
install-node-deps.yml - this will set up the Prometheus node_exporter and install package dependencies.
prepare-data-dir.yml - this will create any RAID devices required and format devices as XFS. Note: This playbook looks for devices presented to the operating system as NVMe devices (which can include EBS volumes built on the Nitro System). You may replace this playbook with your own method of formatting devices and presenting disks.
If managing TLS with the Redpanda playbooks:
generate-csrs.yml - will create private key and CSR and bring the CSR back to the Ansible host.
If using the Redpanda provided CA: issue-certs.yml - signs the CSR and issues a certificate.
install-certs.yml - Installs the certificate and also applies the redpanda_broker role to the cluster nodes. Note: This will install and start Redpanda (and restart any brokers that do not have skip_node=true set)

TASK [redpanda_broker : Restart Redpanda] *****************************************************************************************************
fatal: [54.214.225.246]: FAILED! => {"msg": "The conditional check '(\"true\" in restart_required_rc.stdout or (is_initialized and (nodeconfig_result.changed or \"Removed\" in package_result.results or \"1 upgraded\" in package_result.results))) and (restart_node | default(\"true\") | bool)' failed. The error was: error while evaluating conditional ((\"true\" in restart_required_rc.stdout or (is_initialized and (nodeconfig_result.changed or \"Removed\" in package_result.results or \"1 upgraded\" in package_result.results))) and (restart_node | default(\"true\") | bool)): 'dict object' has no attribute 'results'\n\nThe error appears to be in '/Users/weswagner/git/tristan/deployment-automation/ansible/playbooks/roles/redpanda_broker/tasks/start-redpanda.yml': line 107, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Restart Redpanda\n  ^ here\n"}

WesWWagner reviewed (approved):
I have done a more thorough code review and am comfortable that this code is more stable and does not introduce any new regressions that don't already exist in the mainline.

andrwng reviewed (commented):
Just chiming in that the node ID change looks good.

Also, longer term, as we polish this repo, is there anything (ticket, epic, etc) tracking introducing some smoke tests to the CI pipeline? It'd be nice to e.g. start up a three-node cluster and running some basic checks against it.
